src/frontend/containers.rs

```diff
@@ -0,0 +1,71 @@
+use std::collections::HashMap;
+
+use super::Value;
+use crate::middleware::{
+    containers::{
+        Array as MiddlewareArray, Dictionary as MiddlewareDictionary, Set as MiddlewareSet,
+    },
+    hash_str, Value as MiddlewareValue,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Set(Vec<Value>, MiddlewareSet);
+
+impl Set {
+  pub fn new(values: Vec<Value>) -> Self {
+    let set = MiddlewareSet::new(&values.iter().map(|v| MiddlewareValue::from(v)).collect()).unwrap();
+    Self(values, set)
+  }
+
+  pub fn middleware_set(&self) -> &MiddlewareSet {
+    &self.1
+  }
+
+  pub fn values(&self) -> &Vec<Value> {
+    &self.0
+  }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Dictionary(HashMap<String, Value>, MiddlewareDictionary);
+
+impl Dictionary {
+    pub fn new(values: HashMap<String, Value>) -> Self {
+        let dict = MiddlewareDictionary::new(
+            &values
+                .iter()
+                .map(|(k, v)| (hash_str(k), MiddlewareValue::from(v)))
+                .collect::<HashMap<_, _>>(),
+        )
+        .unwrap();
+        Self(values, dict)
+    }
+
+    pub fn middleware_dict(&self) -> &MiddlewareDictionary {
+        &self.1
+    }
+
+    pub fn values(&self) -> &HashMap<String, Value> {
+        &self.0
+    }
```

Comment:
If I understand this correctly, it is storing the same `values` in the `FrontendSet` and in its `MiddlewareSet` sub-entry.
Maybe we can have the `FrontendSet` as a direct wrapper to the `MiddlewareSet` without duplicating the `values`? something like
```rust
// in the frontend
pub struct Set(MiddlewareSet);
```
and then for getting the values at the `set.values()` method, internally uses
```rust
pub fn values(&self) -> &Vec<Value> {
    &self.0.iter().collect()
}
```
and in this way there is no need to store the values at the frontend's `Set`.

Reply:
(this last `self.0.iter().collect()` internally uses the `MerkleTree` iterator, which iterates over the leafs of the tree only (not the intermediate nodes))

Reply:
Right, although this (the hashing of the entries) seems to happen only at the `Set` type (not at `Dictionary`&`Array`, which can recover the original values from the underlying `MerkleTree`. I guess that we could still have `Dictionary` & `Array` as a direct wrapper on the middleware respective types, and the `Set` being the only one that stores the `values` at the frontend type; but for the current version probably is preferable to have the 3 types (`Dictionary`,`Set`,`Array`) with the same behaviour so that we don't have different logic between them.
So I will mark this comment as 'resolved' and we can keep it as you did it.